File: EvaluationService/src/sandbox/workspace.ts
import fs from "fs";
import path from "path";
import crypto from "crypto";

export interface Workspace {
  id: string;
  dir: string;
}

export function createWorkspace(): Workspace {
  const id = crypto.randomUUID();
  const dir = path.join(process.cwd(), "temp_workspaces", `judge-${id}`);
  fs.mkdirSync(dir, { recursive: true });
  return { id, dir };
}

export function writeFile(
  workspace: Workspace,
  filename: string,
  content: string
) {
  fs.writeFileSync(path.join(workspace.dir, filename), content);
}

export function readFile(workspace: Workspace, filename: string): string {
  return fs.readFileSync(path.join(workspace.dir, filename), "utf8");
}

export function cleanupWorkspace(workspace: Workspace) {
  fs.rmSync(workspace.dir, { recursive: true, force: true });
}

File: evaluation-service-go/internal/sandbox/workspace.go
package sandbox

import (
	"fmt"
	"os"
	"path/filepath"
)

type Workspace struct {
	Dir string
}

func CreateWorkspace() (*Workspace, error) {
	dir, err := os.MkdirTemp("", "judge-*")
	if err != nil {
		return nil, fmt.Errorf("failed to create workspace: %w", err)
	}
	return &Workspace{Dir: dir}, nil
}

func WriteFile(ws *Workspace, filename, content string) error {
	path := filepath.Join(ws.Dir, filename)
	return os.WriteFile(path, []byte(content), 0644)
}

func CleanupWorkspace(ws *Workspace) {
	if ws.Dir != "" {
		os.RemoveAll(ws.Dir)
	}
}
